File: Venkatesh/Merge-Sort/MergeSort.py
def merge_sort(array, n, m):
    sorted = array[:n]
    unsorted = array[n: n+m]

    def merge_sort_recursion(array):
        if len(array) > 1:
            mid = len(array) // 2
            left = array[:mid]
            right = array[mid:]

            merge_sort_recursion(left)
            merge_sort_recursion(right)

            i = j = k = 0
            while i < len(left) and j < len(right):
                if left[i] < right [j]:
                    array[k] = left[i]
                    i += 1
                else:
                    array[k] = right[j]
                    j += 1
                k += 1
            
            while i < len(left):
                array[k] = left[i]
                i += 1
                k += 1

            while j < len(right):
                array[k] = right[j]
                j += 1
                k += 1
    
    merge_sort_recursion(unsorted)

    def merge(sorted, unsorted):
        result = []
        while sorted and unsorted:
            if sorted[0] < unsorted[0]:
                result.append(sorted.pop(0))
            else:
                result.append(unsorted.pop(0))
        result.extend(sorted or unsorted)
        return result

    sorted_array = merge(sorted, unsorted)
    return sorted_array


arr = [1, 3, 5, 9, 84, 26, 62, 12, 71, 20]
n = 4
m = 6

print(merge_sort(arr, n, m))

File: Venkatesh/Merge-Sort/MergeSort.java
public class MergeSort {

    public static void mergeSort(int[] array, int n, int m) {
        int[] sorted = new int[n];
        int[] unsorted = new int[m];
        System.arraycopy(array, 0, sorted, 0, n);
        System.arraycopy(array, n, unsorted, 0, m);

        mergeSortRecursion(unsorted, 0, m - 1);
        merge(sorted, unsorted, array);
    }

    private static void mergeSortRecursion(int[] array, int left, int right) {
        if (left < right) {
            int mid = (left + right) / 2;

            mergeSortRecursion(array, left, mid);
            mergeSortRecursion(array, mid + 1, right);

            merging(array, left, mid, right);
        }
    }

    private static void merging(int[] array, int left, int mid, int right) {
        int n1 = mid - left + 1;
        int n2 = right - mid;

        int[] L = new int[n1];
        int[] R = new int[n2];

        for (int i = 0; i < n1; i++) {
            L[i] = array[left + i];
        }
        for (int j = 0; j < n2; j++) {
            R[j] = array[mid + 1 + j];
        }

        int i = 0, j = 0;
        int k = left;
        while (i<n1 && j<n2) {
            if (L[i] <= R[j]) {
                array[k] = L[i];
                i++;
            } else {
                array[k] = R[j];
                j++;
            }
            k++;
        }
        while (i < n1) {
            array[k] = L[i];
            i++;
            k++;
        }
        while (j < n2) {
            array[k] = R[j];
            j++;
            k++;
        }
    }

    private static void merge(int[] sorted, int[] unsorted, int[] array) {
        int i = 0, j = 0, k = 0;
        while (i < sorted.length && j < unsorted.length) {
            if (sorted[i] < unsorted[j]) {
                array[k++] = sorted[i++];
            } else {
                array[k++] = unsorted[j++];
            }
        }
        while (i < sorted.length) {
            array[k++] = sorted[i++];
        }
        while (j < unsorted.length) {
            array[k++] = unsorted[j++];
        }
    }
    
    public static void main(String[] args) {
        int[] arr = {1, 5, 9, 11, 15, 19, 77, 30, 80, 97};
        int n = 6;
        int m = 4;
        mergeSort(arr, n, m);
        for (int i : arr) {
            System.out.print(i + ", ");
        }
    }
}